cogs/roles.py

```diff
@@ -248,246 +246,40 @@ async def create_role(self, ctx, *, role: Optional[str] = None):
-                continue
-            try:
-                await user.remove_roles(role, reason=f"{ctx.author} put {user} in the penalty box")
-            except (discord.Forbidden, discord.HTTPException):
-                failed_roles.append(str(role))
-
-        # Add the penalty role to the user
-        await user.add_roles(penalty_role, reason=reason_message)
-        await ctx.send(reason_message)
-
-        if failed_roles:
-            await ctx.send(
-                f"The following role{'s' if len(failed_roles) > 0 else ''} could not be removed: {', '.join(failed_roles)}"
-            )
-
-    @commands.command(aliases=["unpenalty"])
-    @is_moderator()
-    async def unmute(self, ctx, user: discord.Member):
-        penalty_role = utils.get(ctx.guild.roles, name=self.penalty_role)
-
-        if penalty_role is None:
-            self.bot.dev_logger.error(PENALTY_ROLE_ERROR)
-            await ctx.send(PENALTY_ROLE_ERROR)
-            return
-
-        if self._is_in_penalty_table(user) and not self._has_penalty_role(user):
-            # User had penalty role manually removed already
-            self._remove_from_penalty_table(user)
-            await ctx.send(f"{user.display_name} was already unmuted manually;" f" removed role history from database")
+        if is_in_muted_table(self, user):
+            await ctx.send("Cannot restore roles to a muted user")
             return
 
-        reason_message = f"{ctx.author} removed {user} from the penalty box"
-
-        # Restore old roles from the database
-        valid_roles = self._fetch_saved_roles(ctx.guild, user, penalty=True)
-        await self._role_restoring_page(ctx, user, valid_roles)
-
-        # Remove the penalty role
-        await user.remove_roles(penalty_role, reason=reason_message)
-
-        # Remove entry from the database
-        self._remove_from_penalty_table(user)
-
-        await ctx.send(reason_message)
+        valid_roles = fetch_saved_roles(self, ctx.guild, user)
+        await role_restoring_page(self, ctx, user, valid_roles)
 
-    @commands.command(aliases=["inchannel"])
-    async def in_channel(self, ctx):
-        """Returns list of users in current channel"""
-        channel = ctx.message.channel
-        members = channel.members
-
-        channel_users = list(map(lambda m: str(m) + "\n", members))
-        header = f"List of users in #{channel} - {len(members)}"
-
-        pages = Pages(ctx, item_list=channel_users, title=header, display_option=(3, 20), editable_content=False)
-        await pages.paginate()
+    @commands.Cog.listener()
+    async def on_member_remove(self, user: discord.Member):
+        if is_in_muted_table(self, user):
+            # Check if the user is muted.
+            # If so, save all roles BUT the muted role into the PreviousRoles table
+            save_existing_roles(self, user, muted=True)
+        else:
+            # Save existing roles
+            save_existing_roles(self, user)
```

Comment:
can this maybe be factored into
```py
@commands.Cog.listener()
async def on_member_remove(self, user: discord.Member):
    save_existing_roles(self, user, muted=is_in_muted_table(self, user))
```

also passing self as an explicit param is sort of non-conventional; you could make a mixin class and use multiple inheritance, or pass in the DB path or the bot object instead perhaps

Reply:
resolved in 8af4976